internal/eru/recycle/recycle.go

```diff
@@ -0,0 +1,153 @@
+	// so there is a time window in which the guest is a dangling guest, so we wait for a while and wait the deletion finished
+	// TODO better way to detect if a guest is in deletion
+	time.Sleep(deleteWait)
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer cancel()
+	logger.Infof(ctx, "[recycle] start to remove dangling guest %s", eruID)
+	// since guest deletion is a dangerous operation here,
+	// so we check eru again
+	wrk, err := sto.GetWorkload(ctx, eruID)
+	logger.Infof(ctx, "[recycle]  guest %s, wrk: %v, err: %v", eruID, wrk, err)
+	if err == nil {
+		logger.Errorf(ctx, err, "[recycle] BUG: dangling guest %s is still in eru", eruID)
+		return errors.Errorf("BUG: dangling guest %s is still in eru", eruID)
+	}
+
+	e, ok := status.FromError(err)
+	if !ok {
+		return err
+	}
+	if e.Code() == 1051 && strings.Contains(e.Message(), "entity count invalid") { //nolint
+		logger.Infof(ctx, "[recycle] start to remove local guest %s", eruID)
+		// When creating a guest, the core first creates the workload and then creates a record in ETCD.
+		// Therefore, within the time window between these two operations, we may incorrectly detect dangling guests.
+		// To prevent this situation, we create a creation session locker when creating a guest and check this locker here.
+		flck := utils.NewCreateSessionFlock(utils.VirtID(eruID))
+		if flck.FileExists() {
+			// creation session locker file exists
+			// it means this guest is in creation
+			logger.Warnf(ctx, "[recycle] guest %s in creation", eruID)
+			return fmt.Errorf("guest %s is in creation", eruID)
+		}
+
+		if err := svc.ControlGuest(ctx, utils.VirtID(eruID), virttypes.OpDestroy, true); err != nil {
+			logger.Errorf(ctx, err, "[recycle] failed to remove dangling guest %s", eruID)
+			return err
+		}
+		notifier := bison.GetService()
+		log.Debugf(ctx, "[recycle] notifier: %v", notifier)
+		if notifier != nil {
+			msgList := []string{
+				"<font color=#00CC33 size=10>delete dangling successfully </font>",
+				"---",
+				"",
+				fmt.Sprintf("- **node:** %s", configs.Hostname()),
+				fmt.Sprintf("- **id:** %s", eruID),
+			}
+			text := "\n" + strings.Join(msgList, "\n")
+			if err := notifier.SendMarkdown(context.TODO(), "delete dangling guest", text); err != nil {
+				logger.Warnf(ctx, "[recycle] failed to send dingtalk message: %v", err)
+			}
+		}
+		return nil
+	}
+
+	return err
+}
+
+func startLoop(ctx context.Context, svc service.Service) {
+	log.WithFunc("startLoop").Infof(ctx, "[recycle] starting recycle loop")
+	defer log.WithFunc("startLoop").Infof(ctx, "[recycle] recycle loop stopped")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+		}
+
+		coreIDs, err := fetchWorkloads()
+		if err != nil {
+			continue
+		}
+		localIDs, err := svc.GetGuestIDList(context.Background())
+		if err != nil {
+			continue
+		}
+		coreMap := map[string]struct{}{}
+		for _, id := range coreIDs {
```

Comment:
can we utilize the `lo.Map`

Reply:
lo.Map returns slice and here is a map and i think for-loop is more intuitive here